src/main/studio/ui/ServerList.java

```diff
@@ -488,31 +561,125 @@ private void addNode(boolean folder, AddNodeLocation location) {
             server.setFolder(parent);
             newNode = new ServerTreeNode(server);
         }
+        addExistingNode(node, newNode, location);
+    }
 
-        int index;
-        if (location == AddNodeLocation.INSERT) {
-            index = node.getChildCount();
-        } else {
-            index = parent.getIndex(node);
-            if (location == AddNodeLocation.AFTER) index++;
+    private void pickUpNode() {
+        pickedUpNode = (ServerTreeNode) tree.getLastSelectedPathComponent();
+    }
+    private void dropNode(AddNodeLocation location) {
+        if (pickedUpNode == null) return;
+        ServerTreeNode selNode  = (ServerTreeNode) tree.getLastSelectedPathComponent();
+        if (selNode == pickedUpNode) {
+            JOptionPane.showMessageDialog(this,"Cannot move a node relative to itself.","Error",JOptionPane.ERROR_MESSAGE,Util.ERROR_ICON);
+            return;
         }
-
-        parent.insert(newNode, index);
-        try {
-            Config.getInstance().setServerTree(serverTree);
-        } catch (IllegalArgumentException exception) {
-            serverTree = Config.getInstance().getServerTree();
-            log.error("Error adding new node", exception);
-            StudioOptionPane.showError(this, "Error adding new node:\n" + exception, "Error");
+        if (pickedUpNode.isFolder()) {
+            ServerTreeNode dropTarget = location == AddNodeLocation.INSERT ? selNode : (ServerTreeNode) selNode.getParent();
+            while (dropTarget != root) {
+                if (dropTarget == pickedUpNode) {
+                    JOptionPane.showMessageDialog(this,"Cannot move a folder into itself.","Error",JOptionPane.ERROR_MESSAGE,Util.ERROR_ICON);
+                    return;
+                }
+                dropTarget = (ServerTreeNode) dropTarget.getParent();
+            }
         }
-        refreshServers();
+        addExistingNode(selNode, pickedUpNode, location);
+    }
```

Comment:
From UX perspective:
* I would implement D&D
* If we implement pick-up solution, I would have visual indication of the picked up item (font/icon)
* There is Server/Import servers from QPad menu item. IMHO I would put import/export json somewhere close into the main menu
* When I picked and moved a server, comboServer didn't refresh. And as a result I wasn't able to select the server from the comboBox

Reply:
#82 created for drag and drop as it wouldn't fit the scope of this PR.

Reply:
I tried to change the label of the tree node when a node is picked up but looks like how things behave doesn't reflect what is in the docs. So I'll leave that for now. If drag and drop is implemented this would be unnecessary anyway.